crates/driver/src/infra/simulator/mod.rs

```diff
@@ -78,6 +90,7 @@ impl Simulator {
                     .gas
             }
             Inner::Ethereum(ethereum) => ethereum.estimate_gas(tx).await?,
+            Inner::Enso(enso, _) => enso.simulate(tx).await?,
         })
```

Comment:
To me, there are currently two gas estimates we are interested in:
1. `gas_estimate`
2. `gas_used`

with difference between these explained with an example. Let's say our transaction spends 150k gas, but at the end of execution in EVM it gets 20k gas refunds, meaning it will spend 130k gas. User still has to have 150k gas available to pay in his account so the EVM does not return `out_of_gas` error.
In this example, `gas_estimate` would be >=150k while the `gas_used` would be 130k gas.

We would be interested in `gas_estimate` to check the user account balances etc, but also interested in `gas_used` to calculate expected costs etc.

I would suggest to:
1. Split `gas` function into two functions, `gas_estimate` and `gas_used`.
2. Allow `Simulator` to work with `Vec<Inner>` instead of `Inner`.

If we want to do this it can be done in a separate PR.

Reply:
Agree that these are two different use cases. 
Currently we interpret gas in the way you described `gas_used` and then add 20% buffer to arrive at a reasonable gas limit. I think this may be ok for simplicity, given the current resources we have. We could simply increase the buffer if needed (you never really spend the gas limit unless you are a bit careless and run into into invalid() reverts). 

If we make gas estimates a first class citizen, I wonder if Simulator is the right abstraction for it? Isn't it something so basic to the network that it should be part of Eth (the same could potentially be argued about access lists).

Reply:
I agree, we can definitely leave it as is and reevaluate after.

Reply:
I also think that gas estimates should be part of `Ethereum` and not the simulator. In fact, it would be nice to change the code to reflect this.